### PR TITLE
Improve the documentation for flatAlt

### LINKED ISSUE #151: Confusing documentation for flatAlt
https://github.com/quchen/prettyprinter/blob/1996a1692c1d32418626d3af0ba084a71892ed21/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L159-L162

https://github.com/quchen/prettyprinter/blob/1996a1692c1d32418626d3af0ba084a71892ed21/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L671-L707

To me, the word "fallback" suggests something less desirable than the (preferred) alternative. In these docs "fallback" is used to refer to the second `FlatAlt` field, which however is the one we _prefer_ when `group`ed.

### PR DESCRIPTION
**Pull request by sjakobi**
Closes #151.

### CODE REVIEW

File: prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs [hunk 1]
```diff
@@ -668,10 +668,12 @@ changesUponFlattening = \doc -> case doc of
 
 
 
--- | @('flatAlt' x fallback)@ renders as @x@ by default, but falls back to
--- @fallback@ when 'group'ed. Since the layout algorithms rely on 'group' having
--- an effect of shortening the width of the contained text, careless usage of
--- 'flatAlt' with wide fallbacks might lead to unappealingly long lines.
+-- | By default, @('flatAlt' x y)@ renders as @x@. However when 'group'ed,
+-- @y@ will be preferred, with @x@ as the fallback for the case where @y@
+-- doesn't fit.
```
Comment (sjakobi) [inline prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs]: I think I should also mention that `group` will (attempt to) flatten `y`.

File: prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs [hunk 2]
```diff
@@ -668,10 +668,12 @@ changesUponFlattening = \doc -> case doc of
...
+--
```
Comment (sjakobi) [inline prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs]: I should add some simple examples here.

File: prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs (hunk 1 above)
Comment (sjakobi) [reply prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs]: …and that `y` will never be chosen if it's unflattenable! (https://github.com/quchen/prettyprinter/issues/152)

**Comment by sjakobi** [comment]
I'm fairly happy with this now, despite the length.

Reviews welcome! :)

**Comment by quchen** [comment]
Looks good!

For longer documentation including extensive examples, we might start using collapsible sections in Haddock, so we have to click on them to reveal the wall of text. Just a thought.

**Comment by sjakobi** [comment]
> For longer documentation including extensive examples, we might start using collapsible sections in Haddock, so we have to click on them to reveal the wall of text.

In fact that's what I used for the do-notation example!

**Pull request merged**